src/Swarm/Language/LSP/Hover.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Swarm.Language.LSP.Hover where
+
+import Data.Rope.UTF16 qualified as R
+import Data.Text (Text)
+import Data.Text qualified as T
+import Language.LSP.Types qualified as J
+import Language.LSP.VFS
+import Swarm.Language.Parse (readTerm')
+import Swarm.Language.Syntax
+
+withinBound :: Int -> SrcLoc -> Bool
+withinBound pos (SrcLoc start end) = pos >= start && pos < end
+withinBound _ NoLoc = False
+
+showHoverInfo ::
+  J.NormalizedUri ->
+  J.TextDocumentVersion ->
+  J.Position ->
+  VirtualFile ->
+  Maybe Text
+showHoverInfo _ _ (J.Position myLine myCol) vf@(VirtualFile _ _ myRope) =
+  astSize
+ where
+  content = virtualFileText vf
+  absolutePos = R.rowColumnCodeUnits (R.RowColumn (fromIntegral myLine) (fromIntegral myCol)) myRope
+  astSize = case readTerm' content of
+    Right Nothing -> Nothing
+    Right (Just (Syntax _ term)) -> Just $ explain $ digAt term absolutePos
+    Left _ -> Nothing
+
+descend ::
+  -- | default
+  Term ->
+  -- | position
+  Int ->
+  -- | next element to inspect
+  Syntax ->
+  Term
+descend t pos (Syntax l1 t1) =
+  if withinBound pos l1
+    then digAt t1 pos
+    else t
+
+descend2 ::
```

Comment:
Is there a nicer way to express this, that does not require defining `descend` and `descend2`?

Reply:
Yes, I like that better.